File: tests/cross_driver/java/src/main/java/com/secantus/smokes/PbrtSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.ChangeStreamIterable;
import com.mongodb.client.MongoChangeStreamCursor;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.changestream.ChangeStreamDocument;
import org.bson.BsonDocument;
import org.bson.Document;

import java.util.concurrent.TimeUnit;

/**
 * Cross-driver postBatchResumeToken smoke — Java.
 * <p>
 * Pin a change-stream cursor before any inserts; subsequent
 * {@code tryNext()} calls pull empty batches but the resume token
 * should still advance — that's what postBatchResumeToken delivers.
 */
public final class PbrtSmoke {

    private PbrtSmoke() {}

    public static void main(String[] args) throws Exception {
        String uri = System.getenv("MONGODB_URI");
        if (uri == null) {
            System.err.println("MONGODB_URI not set");
            System.exit(2);
        }

        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .applyToClusterSettings(b -> b.serverSelectionTimeout(30, TimeUnit.SECONDS))
            .build();

        try (MongoClient client = MongoClients.create(settings)) {
            MongoCollection<Document> coll = client.getDatabase("pbrt_xd").getCollection("c");
            coll.drop();

            ChangeStreamIterable<Document> cs = coll.watch().maxAwaitTime(500, TimeUnit.MILLISECONDS);
            try (MongoChangeStreamCursor<ChangeStreamDocument<Document>> cur = cs.cursor()) {
                String initial = tokenSig(cur.getResumeToken());
                cur.tryNext();
                Thread.sleep(200);
                cur.tryNext();
                Thread.sleep(200);
                cur.tryNext();
                String after = tokenSig(cur.getResumeToken());

                if (after == null || after.isEmpty()) {
                    System.err.printf("no resume token after empty polls (initial=%s)%n", initial);
                    System.exit(1);
                }
                if (!initial.isEmpty() && initial.equals(after)) {
                    System.err.printf("resume token did not advance across empty getMores: %s%n", after);
                    System.exit(1);
                }
                System.out.println("OK");
            }
        }
    }

    private static String tokenSig(BsonDocument token) {
        if (token == null) {
            return "";
        }
        if (token.containsKey("_data") && token.get("_data").isString()) {
            return token.getString("_data").getValue();
        }
        return token.toJson();
    }
}

File: tests/cross_driver/node/pbrt_smoke.js
// Cross-driver postBatchResumeToken smoke — Node.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function tokenSig(token) {
  if (!token) return '';
  // Resume tokens are `{_data: <hex>}` — compare on _data.
  return typeof token._data === 'string' ? token._data : JSON.stringify(token);
}

async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) fail('MONGODB_URI not set');

  const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });
  await client.connect();
  try {
    const coll = client.db('pbrt_xd').collection('c');
    await coll.drop().catch(() => {});

    const cs = coll.watch([], { maxAwaitTimeMS: 500 });
    try {
      // Pin the cursor before any inserts; subsequent tryNext() pulls
      // empty batches, but the resume token should still advance.
      const initial = tokenSig(cs.resumeToken);
      await cs.tryNext();
      await sleep(200);
      await cs.tryNext();
      await sleep(200);
      await cs.tryNext();
      const after = tokenSig(cs.resumeToken);

      if (!after) fail(`no resume token after empty polls (initial=${JSON.stringify(initial)})`);
      if (initial && initial === after) {
        fail(`resume token did not advance across empty getMores: ${after}`);
      }
      console.log('OK');
    } finally {
      await cs.close();
    }
  } finally {
    await client.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));
